styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -346,76 +348,78 @@ public void cleanup(WorkflowInstance workflowInstance, String executionId) {
-          }
-        }));
+  void cleanupWithoutRunState(WorkflowInstance workflowInstance, Pod pod) {
+    final Optional<ContainerStatus> containerStatus = getMainContainerStatus(pod);
+    if (!containerStatus.isPresent()) {
+      deletePod(workflowInstance, pod, "No RunState, no container status");
+    } else if (isTerminated(containerStatus.get())) {
+      deletePodIfNonDeletePeriodExpired(workflowInstance, pod, containerStatus.get());
+    } else {
+      // Refresh the pod if it is not terminated as it might be stale
+      pod = client.pods().withName(pod.getMetadata().getName()).get();
+      if (pod == null) {
+        return;
+      }
+      // if not terminated, delete directly
+      if (!isTerminated(pod)) {
+        deletePod(workflowInstance, pod, "No RunState, not terminated");
+      }
+    }
   }
 
   private void deletePodIfNonDeletePeriodExpired(WorkflowInstance workflowInstance,
-                                                 String executionId,
+                                                 Pod pod,
                                                  ContainerStatus containerStatus) {
     if (isNonDeletePeriodExpired(containerStatus)) {
       // if terminated and after graceful period, delete the pod
       // otherwise wait until next polling happens
-      deletePod(workflowInstance, executionId);
+      deletePod(workflowInstance, pod, "Terminated and expired");
     }
   }
 
-  private void cleanup(WorkflowInstance workflowInstance, String executionId,
-                       Consumer<Pod> cleaner) {
-    Optional.ofNullable(client.pods().withName(executionId).get()).ifPresent(pod -> {
-      final List<ContainerStatus> containerStatuses = pod.getStatus().getContainerStatuses();
-      if (!containerStatuses.isEmpty()) {
-        cleaner.accept(pod);
-      } else {
-        // for some cases such as evicted pod, there is no container status, so we delete directly
-        deletePod(workflowInstance, executionId);
-      }
-    });
-  }
-
   static Optional<ContainerStatus> getMainContainerStatus(Pod pod) {
     return readPodWorkflowInstance(pod)
         .flatMap(wfi -> pod.getStatus().getContainerStatuses().stream()
             .filter(status -> mainContainerName(pod.getMetadata().getName()).equals(status.getName()))
             .findAny());
   }
 
-  private boolean isNonDeletePeriodExpired(ContainerStatus containerStatus) {
-    return Optional.ofNullable(containerStatus.getState().getTerminated().getFinishedAt())
-        .map(finishedAt -> Instant.parse(finishedAt)
-            .isBefore(time.get().minus(
-                Duration.ofSeconds(podDeletionDelaySeconds))))
-        .orElse(true);
+  private boolean isNonDeletePeriodExpired(ContainerStatus cs) {
+    final ContainerStateTerminated t = cs.getState().getTerminated();
+    if (t.getFinishedAt() == null) {
+      return true;
+    }
+    final Instant finishedAt;
+    try {
+      finishedAt = Instant.parse(t.getFinishedAt());
```

Comment:
To be paranoid?

Reply:
Always assume data from external systems might be corrupt.